#!/usr/bin/env Rscript

library(rmarkdown)
library(optparse)


option_list = list(
    make_option(c("-r", "--report"), type="character", default=NULL, help="report template file", metavar="character"),
    make_option(c("-o", "--output"), type="character", default="ampliseq_report.html", help="output file name", metavar="character"),
    make_option(c("--skip_fastqc"), action="store_true", default=FALSE, help="Trigger to skip fastqc reporting", metavar="logical"),
    make_option(c("--skip_cutadapt"), action="store_true", default=FALSE, help="Trigger to skip cutadapt filtering", metavar="logical"),
    make_option(c("--skip_dada_quality"), action="store_true", default=FALSE, help="Trigger to skip dada2 quality plotting", metavar="logical"),
    make_option(c("--skip_barrnap"), action="store_true", default=FALSE, help="Trigger to skip barrnap ASV filtering", metavar="logical"),
    #make_option(c("--skip_taxonomy"), action="store_true", default=FALSE, help="Trigger to skip taxonomic classification", metavar="logical"),
    make_option(c("--retain_untrimmed"), action="store_true", default=FALSE, help="Flag to retain the untrimmed sequences", metavar="logical"),
    make_option(c("--ref_tax_user"), action="store_true", default=FALSE, help="Flag that user provided custom db", metavar="logical"),
    make_option(c("--single_end"), action="store_true", default=FALSE, help="Flag if single end data is used", metavar="logical"),
    make_option(c("--trunclenf"), type="numeric", default=-1, help="Parameter to define truncation in forward strand", metavar="numeric"),
    make_option(c("--trunclenr"), type="numeric", default=-1, help="Parameter to define truncation in reverse strand", metavar="numeric"),
    make_option(c("--max_ee"), type="numeric", default=-1, help="Parameter to filter reads based on expected errors", metavar="numeric"),
    make_option(c("--trunc_qmin"), type="numeric", default=-1, help="Parameter to define truncation via quality measure. Set to -1 if trunclen were given.", metavar="numeric"),
    make_option(c("--trunc_rmin"), type="numeric", default=-1, help="Parameter to define truncation via read retaining ratio. Set to -1 if trunclen were given.", metavar="numeric"),
    make_option(c("--mqc_plot"), type="character", default=NULL, help="MultiQC plot per sequence quality", metavar="character"),
    make_option(c("--ca_sum_path"), type="character", default=NULL, help="cutadapt summary table", metavar="character"),
    make_option(c("--dada_filtntrim_args"), type="character", default=NULL, help="DADA2 arguments for filter and trim process", metavar="character"),
    make_option(c("--dada_qc_f_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_qc_r_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_pp_qc_f_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_pp_qc_r_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_1_err_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_2_err_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--asv_table_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_asv_fa"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_dada2_tab"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada_stats_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_rrna_arc"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_rrna_bac"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_rrna_euk"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_rrna_mito"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--path_barrnap_sum"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--filter_len_asv"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--min_len_asv"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--max_len_asv"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--filter_codons"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--stop_codons"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--filter_len_asv_len_orig"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--ref_tax_path"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--dada2_taxonomy"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--flag_dada2_taxonomy"), action="store_true", default=FALSE, help="MultiQC plots", metavar="character"),
    make_option(c("--sintax_taxonomy"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--flag_sintax_taxonomy"), action="store_true", default=FALSE, help="MultiQC plots", metavar="character"),
    make_option(c("--pplace_taxonomy"), type="character", default=NULL, help="MultiQC plots", metavar="character"),
    make_option(c("--flag_pplace_taxonomy"), action="store_true", default=FALSE, help="MultiQC plots", metavar="character"),
    make_option(c("--flag_qiime2_taxonomy"), action="store_true", default=FALSE, help="MultiQC plots", metavar="character"),
    make_option(c("--qiime2_taxonomy"), type="character", default=NULL, help="MultiQC plots", metavar="character")
)

opt_parser = OptionParser(option_list = option_list)
opt = parse_args(opt_parser)

rmarkdown::render(opt$report, output_file = opt$output,
                    params = list(
                        flag_skip_fastqc = opt$skip_fastqc,
                        flag_skip_cutadapt = opt$skip_cutadapt,
                        flag_skip_dada_quality = opt$skip_dada_quality,
                        flag_skip_barrnap = opt$skip_barrnap,
                        #flag_skip_taxonomy = opt$skip_taxonomy,
                        flag_retain_untrimmed = opt$retain_untrimmed,
                        flag_ref_tax_user = opt$ref_tax_user,
                        flag_single_end = opt$single_end,
                        trunclenf = opt$trunclenf,
                        trunclenr = opt$trunclenr,
                        max_ee = opt$max_ee,
                        trunc_qmin = opt$trunc_qmin,
                        trunc_rmin = opt$trunc_rmin,
                        mqc_plot = opt$mqc_plot,
                        ca_sum_path = opt$ca_sum_path,
                        dada_filtntrim_args = opt$dada_filtntrim_args,
                        dada_qc_f_path = opt$dada_qc_f_path,
                        dada_qc_r_path = opt$dada_qc_r_path,
                        dada_pp_qc_f_path = opt$dada_pp_qc_f_path,
                        dada_pp_qc_r_path = opt$dada_pp_qc_r_path,
                        dada_1_err_path = opt$dada_1_err_path,
                        dada_2_err_path = opt$dada_2_err_path,
                        asv_table_path = opt$asv_table_path,
                        path_asv_fa = opt$path_asv_fa,
                        path_dada2_tab = opt$path_dada2_tab,
                        dada_stats_path = opt$dada_stats_path,
                        path_rrna_arc = opt$path_rrna_arc,
                        path_rrna_bac = opt$path_rrna_bac,
                        path_rrna_euk = opt$path_rrna_euk,
                        path_rrna_mito = opt$path_rrna_mito,
                        path_barrnap_sum = opt$path_barrnap_sum,
                        filter_len_asv = opt$filter_len_asv,
                        min_len_asv = opt$min_len_asv,
                        max_len_asv = opt$max_len_asv,
                        filter_len_asv_len_orig = opt$filter_len_asv_len_orig,
                        filter_codons = opt$filter_codons,
                        stop_codons = opt$stop_codons,
                        ref_tax_path = opt$ref_tax_path,
                        dada2_taxonomy = opt$dada2_taxonomy,
                        flag_dada2_taxonomy = opt$flag_dada2_taxonomy,
                        flag_sintax_taxonomy = opt$flag_sintax_taxonomy,
                        sintax_taxonomy = opt$sintax_taxonomy,
                        flag_pplace_taxonomy = opt$flag_pplace_taxonomy,
                        pplace_taxonomy = opt$pplace_taxonomy,
                        flag_qiime2_taxonomy = opt$flag_qiime2_taxonomy,
                        qiime2_taxonomy = opt$qiime2_taxonomy))
